### ISSUE TITLE
sql: explicit SELECT FOR UPDATE should be applied after predicate

### ISSUE THREAD

**Issue by nvb** [cockroachdb/cockroach#75457]
**Describe the problem**

This is similar to https://github.com/cockroachdb/cockroach/issues/57031, but it's not the same. Currently, `SELECT FOR UPDATE` places locks on each key scanned by the base index scan. This means that even if some of those keys are later filtered out by a predicate which could not be pushed into the scan, they will still be locked.

This is not the correct behavior (according to PG), but was initially implemented for ease of implementation and to avoid a two-phase locking process with multiple RPCs.

**To Reproduce**

Setup
```sql
create table t(a int primary key, b bool);
insert into t values (1, false), (2, true);
```

Example
```sql
--tx1                                          --tx2
begin transaction;
                                               begin transaction;
select * from t  
where b=true for update;
                                               select * from 
                                               t where b=false for update; -- BLOCKS!
```

**Expected behavior**

txn2's `SELECT` query should not block, because txn1 should have only acquire a lock on the row that it returned (`(2, true)`).

Jira issue: CRDB-12682

Epic CRDB-25322

**Comment by yuzefovich** [comment]
@nvanbenschoten what's the urgency on this?

**Comment by rytaft** [comment]
Moving this to the backlog for now, @nvanbenschoten please let us know if there is urgency on this. Thanks!

**Comment by nvanbenschoten** [comment]
This is also the kind of issue that won't be urgent until it suddenly is, so it's tough to say. The effect of this is that an application or tool built for PG that uses `SELECT FOR UPDATE` could see unexpected contention/transaction deadlocks when run on CockroachDB, because CRDB grabs too many row-level locks in some cases due to this issue. The only mitigations right now would be to remove the `FOR UPDATE` clause or to add secondary indexes to ensure that all predicates are pushed into scans, but this second options runs into issues with #57031.

**Comment by rytaft** [comment]
Ok makes sense -- moving this to a higher priority bucket -- we'll try to get to this during stability if not before

**Comment by nvanbenschoten** [comment]
I suspect that this is too large for stability as it's not trivial, so we may just want to consider it for the v22.2 release cycle.

**Comment by rytaft** [comment]
sounds good -- cc @vy-ton for awareness

**Comment by michae2** [comment]
One of the things we'll have to address when fixing this: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/exec/execbuilder/mutation.go#L1002-L1008

**Comment by msirek** [comment]
Fixed by https://github.com/cockroachdb/cockroach/pull/111662

**Comment by michae2** [comment]
This is fixed in 23.2 when using the new SFU implementation (`optimizer_use_lock_op_for_serializable`).
